grpc-client-server/prober_server/main_test.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+	"testing"
+
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func dialer() func(context.Context, string) (net.Conn, error) {
+	listener := bufconn.Listen(1024 * 1024)
+
+	server := grpc.NewServer()
+
+	pb.RegisterProberServer(server, &Server{})
+
+	go func() {
+		if err := server.Serve(listener); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	return func(context.Context, string) (net.Conn, error) {
+		return listener.Dial()
+	}
+}
+
+func TestDoProbes(t *testing.T) {
+	tests := map[string]struct {
+		req       *pb.ProbeRequest
+		err       error
+		failedReq int32
+	}{
+		"simple": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "http://google.com",
+				NumberOfRequests: 1,
+			},
+			failedReq: 0,
+		},
+		"error": {
+			req: &pb.ProbeRequest{
+				Endpoint:         "http://musterror:8080",
+				NumberOfRequests: 3,
+			},
+			failedReq: 3,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			conn, err := grpc.DialContext(context.Background(), "bufnet", grpc.WithContextDialer(dialer()), grpc.WithInsecure())
+			require.NoError(t, err)
+			defer conn.Close()
+
+			client := pb.NewProberClient(conn)
+
+			resp, err := client.DoProbes(context.Background(), tc.req)
+
+			if err == nil {
+				require.NotNil(t, resp.AverageResponseTime)
```

Comment:
Can you think how you could actually test the response time as well?

(hint: you probably don't _actually_ want to connect to Google when doing so!)

Reply:
great suggestion, I have added a test now.

There could be another way of testing it, by extracting the prober and making it take a parameter for time object. I went with a solution where I mock time.Since, as I think it's clearner for this small usecase.

Reply:
The mock `time.Since` works, but the golang HTTP stack has a built-in type for this (just like you used a custom `Dialer`, called an [`http.RoundTripper`](https://pkg.go.dev/net/http#RoundTripper) - try using it :)